lib/models/os.ts

```diff
@@ -0,0 +1,674 @@
+/*
+Copyright 2016 Balena
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import * as Promise from 'bluebird';
+
+import * as bSemver from 'balena-semver';
+import assign = require('lodash/assign');
+import defaults = require('lodash/defaults');
+import find = require('lodash/find');
+import first = require('lodash/first');
+import includes = require('lodash/includes');
+import once = require('lodash/once');
+import reject = require('lodash/reject');
+import * as memoizee from 'memoizee';
+import * as semver from 'semver';
+
+import {
+	deviceTypes as deviceTypesUtil,
+	findCallback,
+	isDevelopmentVersion,
+	notFoundResponse,
+	onlyIf,
+	treatAsMissingApplication,
+} from '../util';
+
+import { hupActionHelper } from '../util/device-actions/os-update/utils';
+
+import { BalenaRequestStreamResult } from '../../typings/balena-request';
+import {
+	DeviceType,
+	ImgConfigOptions,
+	OsUpdateVersions,
+	OsVersions,
+} from '../../typings/balena-sdk';
+import { InjectedDependenciesParam, InjectedOptionsParam } from '../balena';
+
+const BALENAOS_VERSION_REGEX = /v?\d+\.\d+\.\d+(\.rev\d+)?((\-|\+).+)?/;
+
+const DEVICE_TYPES_ENDPOINT_CACHING_INTERVAL = 10 * 60 * 1000; // 10 minutes
+
+const getOsModel = function(
+	deps: InjectedDependenciesParam,
+	opts: InjectedOptionsParam,
+) {
+	const { request, pubsub } = deps;
+	const { apiUrl, isBrowser } = opts;
+
+	const configModel = once(() => require('./config')(deps, opts));
+	const applicationModel = once(() => require('./application')(deps, opts));
+
+	// tslint:disable-next-line:ban-types
+	const withDeviceTypesEndpointCaching = <T extends Function>(fn: T) => {
+		const memoizedFn = memoizee(fn, {
+			maxAge: DEVICE_TYPES_ENDPOINT_CACHING_INTERVAL,
+			primitive: true,
+			promise: true,
+		});
+
+		pubsub.subscribe('auth.keyChange', () => memoizedFn.clear());
+
+		return memoizedFn;
+	};
+
+	/**
+	 * @summary Get device types with caching
+	 * @description Utility method exported for testability.
+	 * @name _getDeviceTypes
+	 * @private
+	 * @function
+	 * @memberof balena.models.os
+	 */
+	const _getDeviceTypes = withDeviceTypesEndpointCaching<
+		() => Promise<DeviceType[]>
```

Comment:
Do we have to explicitly specify this one?

Reply:
Yes, because the config model is in coffeescript, so getDeviceTypes returns `any`.